sdk/src/main/java/io/opentelemetry/sdk/internal/ComponentRegistry.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.internal;
+
+import io.opentelemetry.sdk.common.InstrumentationLibraryInfo;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nullable;
+
+/**
+ * Base class for all the registry classes (Tracer, Meter, etc.).
+ *
+ * @param <V> the type of the registered value.
+ */
+public abstract class ComponentRegistry<V> {
+  private final Object lock = new Object();
+  private final Map<InstrumentationLibraryInfo, V> registry = new ConcurrentHashMap<>();
+
+  /**
+   * Returns the registered value associated with this name and {@code null} version if any,
+   * otherwise creates a new instance and associates it with the given name and {@code null}
+   * version.
+   *
+   * @param instrumentationName the name of the instrumentation library.
+   * @return the registered value associated with this name and {@code null} version.
+   */
+  public V get(String instrumentationName) {
+    return get(instrumentationName, null);
+  }
+
+  /**
+   * Returns the registered value associated with this name and version if any, otherwise creates a
+   * new instance and associates it with the given name and version.
+   *
+   * @param instrumentationName the name of the instrumentation library.
+   * @param instrumentationVersion the version of the instrumentation library.
+   * @return the registered value associated with this name and version.
+   */
+  public V get(String instrumentationName, @Nullable String instrumentationVersion) {
+    InstrumentationLibraryInfo instrumentationLibraryInfo =
+        InstrumentationLibraryInfo.create(instrumentationName, instrumentationVersion);
+    V tracer = registry.get(instrumentationLibraryInfo);
+    if (tracer == null) {
+      synchronized (lock) {
+        // Re-check if the value was added since the previous check, this can happen if multiple
+        // threads try to access the same named tracer during the same time. This way we ensure that
+        // we create only one TracerSdk per name.
+        tracer = registry.get(instrumentationLibraryInfo);
+        if (tracer != null) {
+          // A different thread already added the named Tracer, just reuse.
+          return tracer;
+        }
+        tracer = newComponent(instrumentationLibraryInfo);
+        registry.put(instrumentationLibraryInfo, tracer);
+      }
+    }
```

Comment:
I think all of this code can be replaced with a call to `registry.putIfAbsent(instrumentationLibraryInfo, newComponent(instrumentationLibraryInfo))` if you declare the `registry` as a `ConcurrentMap`.

Reply:
note: You have to `return registry.get(instrumentationLibraryInfo);` as the result. ;)

Reply:
So I'd personally prefer to 

1) Not having to create a new `InstrumentationLibraryInfo` object per `get()` call.
2) Not to do `registry.putIfAbsent(instrumentationLibraryInfo, newComponent(instrumentationLibraryInfo))` as we would also be creating a new component (I wish `putIfAbsent()` took a lambda ;) )

Probably by doing so we would need to do more work, but we would avoid a few allocations ;)

Reply:
Something something java 8!!

The trade-off is with putIfAbsent we don't block. I'd personally prefer non blocking, but that's me. 

Maybe a hybrid where we putIfAbsent if the first one returns null?

Reply:
@carlosalberto if we use the `InstrumentationLibraryInfo` as the key, how do you prevent creation of one? You could create a String key from the pieces, but that ends up creating a StringBuffer instead of the key object. Either way, if you have named tracers like this, then you'll have to create *something* for the key, no?

Reply:
This PR does not change the logic, it just moves it to be shared. I think all observations are good and having this library shared helps us doing the improvements once.

Reply:
That's fair. When this gets merged, I'll put in a quick PR to simplify the logic and make it less blocking.